Refresh FinOps hub upgrade procedure for multi-version FOCUS adoption (#2123)
microsoft/finops-toolkit: docs-mslearn/toolkit/hubs/data-model.md

## Schema version

One of the goals of the FinOps hubs data model is to guarantee backwards compatibility. To support this, each FinOps hub release uses a specific schema version which aligns to a specific FOCUS version. For instance, the `v1_4` schema version aligns with FOCUS 1.4 and the `v1_2` schema with FOCUS 1.2. The **schema version** defines the columns, data types, and allowed values in the tables and functions for each [managed dataset](#managed-datasets-in-finops-hubs).

Data is transformed to the latest supported FOCUS version during ingestion. For instance, if you ingested FOCUS 1.0 data into the `v1_4` schema, it would be converted to FOCUS 1.4 during ingestion. Older data remains in its originally ingested format. For instance, if you previously ingested FOCUS 1.0 data into the `v1_2` schema, it would remain in the `v1_2` format even after upgrading to a newer schema version.

> The functions in the `Hub` database pull from all tables and transform it into a specific version. For instance, `Costs_v1_0` will convert `v1_4` and `v1_2` data back to the `v1_0` format, ensuring backwards compatibility despite upgrading to newer versions. When you're ready to upgrade, you can selectively upgrade each report or integration to use `Costs_v1_2` or `Costs_v1_4` without breaking others. However, if you want a quick, convenience function that always returns the latest version, you can use the unversioned functions, like `Costs` and `Prices`. Unversioned functions are recommended for ad hoc use while versioned functions are recommended for integration points that require a consistent format after upgrades.

Reviewer: Minor: "pull from all tables and transform **it**" -> "**them**" (refers to the tables).

Author: 🤖 [AI][Claude Code] ✅ **Implemented**

Already fixed — data-model.md now reads "pull from all tables and transform **them**" (shipped in 463c2735).

> The functions in the `Hub` database pull from all tables and transform them into a specific version. For instance, `Costs_v1_0` will convert `v1_4` and `v1_2` data back to the `v1_0` format, ensuring backwards compatibility despite upgrading to newer versions. When you're ready to upgrade, you can selectively upgrade each report or integration to use `Costs_v1_2` or `Costs_v1_4` without breaking others. However, if you want a quick, convenience function that always returns the latest version, you can use the unversioned functions, like `Costs` and `Prices`. Unversioned functions are recommended for ad hoc use while versioned functions are recommended for integration points that require a consistent format after upgrades.

Author: ```suggestion
The functions in the `Hub` database pull from all tables and transform them into a specific version. For instance, `Costs_v1_0` will convert `v1_4` and `v1_2` data back to the `v1_0` format, ensuring backwards compatibility despite upgrading to newer versions. When you're ready to upgrade, you can selectively upgrade each report or integration to use `Costs_v1_2` or `Costs_v1_4` without breaking others. However, if you want a quick, convenience function that always returns the latest version, you can use the unversioned functions, like `Costs` and `Prices`. Unversioned functions are convenient for ad hoc use while versioned functions are recommended for integration points that require a consistent format after upgrades.
```

The following table indicates the schema version for each FinOps hub release and which FOCUS version they align to.

| Release | Schema | FOCUS version |
| ------- | ------ | ------------- |

> | 15+     | `v1_4` | 1.4           |

Reviewer: Forward-dated content in a published Learn page: this adds a `15+ -> v1_4 -> 1.4` row (and `v1_4`/`Costs_v1_4` examples on line 55) before 1.4 support has shipped. If docs-mslearn publishes on merge rather than gated to the 1.4 release, this page would publicly state 1.4 is supported ahead of release. Consider holding the `v1_4` row/examples for a later phase, or confirm the publish cadence makes this safe.

Author: This is in a feature branch. If v15 ships without this and we end up in v16, we'll just update them. We have to put something in for now.

One of the goals of the FinOps hubs data model is to guarantee backwards compatibility. To support this, each FinOps hub release uses a specific schema version which aligns to a specific FOCUS version. The **schema version** defines the columns, data types, and allowed values in the tables and functions for each [managed dataset](#managed-datasets-in-finops-hubs).

The hub schema version mirrors the FOCUS specification version it natively ingests. For example, `v1_2` ingests FOCUS 1.2 exports without transformation. Older data ingested under prior schema versions is automatically up-converted by the unversioned functions (`Costs()`, `Prices()`, etc.) so consumers always see the latest schema regardless of when the data was ingested.

The following table indicates the schema version for each FinOps hub release and which FOCUS version they align to.

| Release | Schema | FOCUS version |
| ------- | ------ | ------------- |
| 12+     | `v1_2` | 1.2           |
| 0.7-11  | `v1_0` | 1.0           |

> When hub support ships ahead of a FOCUS specification ratifying or ahead of Cost Management exporting a version, the hub schema is labeled **preview**. Preview schemas may change without notice between releases. The unversioned functions remain pinned to the latest GA schema until the preview promotes to GA, ensuring production reports stay stable while early adopters validate the new version.

Author: 🤖 [AI][Claude Code] ✅ **Implemented**

Reworded to "ahead of FOCUS specification ratification" for parallel structure with the next clause.

> One of the goals of the FinOps hubs data model is to guarantee backwards compatibility. To support this, each FinOps hub release uses a specific schema version which aligns to a specific FOCUS version. The **schema version** defines the columns, data types, and allowed values in the tables and functions for each [managed dataset](#managed-datasets-in-finops-hubs).

Author: ```suggestion
One of the goals of the FinOps hubs data model is to guarantee backwards compatibility. To support this, each FinOps hub release uses a specific schema version which aligns to a specific FOCUS version. For instance, the `v1_4` schema version aligns with FOCUS 1.4 and the `v1_2` schema with FOCUS 1.2. The **schema version** defines the columns, data types, and allowed values in the tables and functions for each [managed dataset](#managed-datasets-in-finops-hubs).
```

> The hub schema version mirrors the FOCUS specification version it natively ingests. For example, `v1_2` ingests FOCUS 1.2 exports without transformation. Older data ingested under prior schema versions is automatically up-converted by the unversioned functions (`Costs()`, `Prices()`, etc.) so consumers always see the latest schema regardless of when the data was ingested.

Author: ```suggestion
Data is transformed to the latest supported FOCUS version during ingestion. For instance, if you ingested FOCUS 1.0 data into the `v1_4` schema, it would be converted to FOCUS 1.4 during ingestion. Older data remains in its originally ingested format. For instance, if you previously ingested FOCUS 1.0 data into the `v1_2` schema, it would remain in the `v1_2` format even after upgrading to a newer schema version.

The functions in the `Hub` database pull from all tables and transform it into a specific version. For instance, `Costs_v1_0` will convert `v1_4` and `v1_2` data back to the `v1_0` format, ensuring backwards compatibility despite upgrading to newer versions. When you're ready to upgrade, you can selectively upgrade each report or integration to use `Costs_v1_2` or `Costs_v1_4` without breaking others. However, if you want a quick, convenience function that always returns the latest version, you can use the unversioned functions, like `Costs` and `Prices`. Unversioned functions are recommended for ad hoc use while versioned functions are recommended for integration points that require a consistent format after upgrades.
```

The following table indicates the schema version for each FinOps hub release and which FOCUS version they align to.

| Release | Schema | FOCUS version |
| ------- | ------ | ------------- |

> | 12+     | `v1_2` | 1.2           |

Author: ```suggestion
| 15+     | `v1_4` | 1.4           |
| 12-14   | `v1_2` | 1.2           |
```

| 0.7-11  | `v1_0` | 1.0           |

When hub support ships ahead of FOCUS specification ratification, the hub schema is labeled **preview**. Preview schemas may change without notice between releases because the underlying spec is still working draft. The unversioned functions remain pinned to the latest GA schema until the preview promotes to GA, ensuring production reports stay stable while early adopters validate the new version.

Cost Management export availability for a given FOCUS version is a separate concern. A hub schema can be GA even if Cost Management has not yet shipped a matching export; in that case, the hub is ready to ingest the new format as soon as the export ships. This is documented in the [hub upgrade procedure](https://github.com/microsoft/finops-toolkit/tree/dev/src/templates/finops-hub/docs/README.md) so adopters know which scenarios require additional setup once an export is available.

> 

Author: ```suggestion
Cost Management export availability for a given FOCUS version is separate from FinOps hub support for that version. FinOps hubs transforms older export formats to the supported schema version (forward or backward, depending on your FinOps hub release).

```
